### ISSUE TITLE
auto-merge functionality

### ISSUE THREAD

**Issue by elray1** [Infectious-Disease-Modeling-Hubs/hubverse-actions#2]
There should be an optional setting at the hub level for whether to auto-merge PRs upon successful validation.

**Comment by elray1** [comment]
I suspect that we would want to implement this as part of something like `hubValidations::validate_pr` rather than as part of an action?

**Comment by annakrystalli** [comment]
Some useful resources here: https://github.com/pascalgn/automerge-action

The above action uses label to decide what to automerge. We could make the validation workflow create an automerge label on the PR. If validation succeeds, it could then trigger the automerge workflow.

Depending on how exactly we implement this, it could be set via hub config or it could be a whole separate GitHub Action that needs to be explicitly added to a hub that is either linked to the validations workflow or just monitors for the automerge label which the validations workflow creates.


It would be good to spell out when exactly this will be activated. for example, if a PR contains changes to files not validated, should auto-merge still be performed if validations pass?

**Comment by annakrystalli** [comment]
Useful resources supplied by @sbidari related to a sandbox hub where the functionality has been implemented 

- https://github.com/CDCgov/covidhub-sandbox
- https://github.com/CDCgov/covidhub-sandbox/blob/main/.github/workflows/merge-validated-pr.yaml
- https://github.com/CDCgov/covidhub-sandbox/blob/main/.github/workflows/validate-submissions.yaml

**Comment by annakrystalli** [comment]
See also related issue https://github.com/hubverse-org/hubValidations/issues/9 about having auto-merge functionality apply to authorised submittersonly which is also implemented globally in the above example. Ideally we'd want each model_id directory to only be editable by the team members specified in the a model metadata property.

**Comment by bsweger** [comment]
Wrote up some thoughts about why the approach linked above is not a good fit for a general-purpose Hubverse GitHub workflow: https://docs.google.com/document/d/1JutZMGtrzsK5BE9M7Yq3Y6_UQWWvmpU4uYJiLvCYoJ8

The next step is determining:

- Whether or not the Hubverse wants to pursue offering this feature
- If so, how we want to implement it

**Comment by bsweger** [comment]
Adding a proposed workflow from @zkamvar 

The basic workflow:

1. Check the list of trusted contributors (GET /contents
2. Check the files in the PR (GET PR/[number]/list-files)
3. If they are all good, send approval (POST) (actual merge would occur via pull request rules around automatic merges)

These are all three API calls that can be run without running untrusted code. 

The `hubValidations` package currently has functionality that [checks pull request contents](https://hubverse-org.github.io/hubValidations/reference/validate_pr.html#details), (implemented [right before standard hubValidations checks](https://github.com/hubverse-org/hubValidations/blob/8c751ea98ee0ded65fb6d1401ceec3e26f54178a/R/validate_pr.R#L140-L187)) which will check for modifications or deletions of any existing model output files. 

The only inputs for this portion of the check are the inputs for the list pr files endpoint: https://docs.github.com/en/rest/pulls/pulls?apiVersion=2022-11-28#list-pull-requests-files.

It could be feasible for this check to exist as a standalone function that could be run in a `pull_request_target` workflow that achieves the same functionality without using third-party actions. That being said, it relies on the assumption that a supply chain attack does not impact `hubValidations` itself.

The POST step is the riskiest part, and multiple workflows are ideal:

1. pull_request_target with read-only permissions saves a "success" artifact if the PR is valid
2. a workflow_run that runs after the pull_request_target with approve permissions that runs only if there is an artifact from the caller workflow. 

This way ensures that repo secrets are protected in case there is a leak in the pull_request_target workflow.

**Comment by bsweger** [comment]
Moving this to "needs clarification" because we need feedback on whether or not this feature (implemented as a version of @zkamvar's suggestion above) should be prioritized.

**Comment by nickreich** [comment]
I don't think this can be prioritized given our current staffing. I'm going to move to wishlist.
